### PR TITLE
Test: check zephyr os tracing functions.

### PR DESCRIPTION
**Pull request by youhuazhu**
Use self-define backends to get tracing data and check
the output data.

Signed-off-by: YouhuaX Zhu <youhuax.zhu@intel.com>

### CODE REVIEW

File: tests/subsys/tracing/prj.conf [hunk 1]
```diff
@@ -0,0 +1 @@
+CONFIG_ZTEST=y
```
Comment (nashif) [inline tests/subsys/tracing/prj.conf:1]: what are we testing here exactly?

File: tests/subsys/tracing/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+
+if(BOARD MATCHES "qemu_.*")
```
Comment (nashif) [inline tests/subsys/tracing/CMakeLists.txt:5]: where is this output being used?

File: tests/subsys/tracing/prj_uart_ctf.conf [hunk 3]
```diff
@@ -0,0 +1,8 @@
+CONFIG_ZTEST=y
```
Comment (nashif) [inline tests/subsys/tracing/prj_uart_ctf.conf:1]: I do not see this file being used anywhere

File: tests/subsys/tracing/testcase.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+tests:
+  tracing.transport:
+    tags: tracing_disabled
```
Comment (nashif) [inline tests/subsys/tracing/testcase.yaml:3]: what is the purpose of this? What are you testing?

File: tests/subsys/tracing/testcase.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  tracing.transport.uart:
+    extra_args: CONF_FILE="prj_uart.conf"
+    filter: CONFIG_UART_CONSOLE and not CONFIG_QEMU_TARGET
+    tags: tracing_uart
+  tracing.transport.usb:
+    extra_args: CONF_FILE="prj_usb.conf"
+    filter: CONFIG_UART_CONSOLE and not CONFIG_QEMU_TARGET
+    platform_exclude: adafruit_feather_m0_basic_proto
```
Comment (nashif) [inline tests/subsys/tracing/testcase.yaml:11]: why is this being excluded, add a comment

File: tests/subsys/tracing/testcase.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+tests:
+  tracing.transport:
+    tags: tracing_disabled
+  tracing.transport.uart:
+    extra_args: CONF_FILE="prj_uart.conf"
+    filter: CONFIG_UART_CONSOLE and not CONFIG_QEMU_TARGET
+    tags: tracing_uart
+  tracing.transport.usb:
+    extra_args: CONF_FILE="prj_usb.conf"
+    filter: CONFIG_UART_CONSOLE and not CONFIG_QEMU_TARGET
```
Comment (nashif) [inline tests/subsys/tracing/testcase.yaml:10]: why is qemu being excluded here? when does this test get run? I see it now only building for all tests where tracing is enabled and only runs on tests that do not enable tracing, so what is the point?

**Review by nashif — CHANGES_REQUESTED** [review]

File: tests/subsys/tracing/prj.conf (hunk 1 above)
Comment (youhuazhu) [reply tests/subsys/tracing/prj.conf:1]: Thanks @nashif ,  This is for the case when tracing functions disabled,   if tracing functions are disabled,  we shouldn't find any tracing data in the backend:
```
//checked in main.c
#ifdef CONFIG_TRACING
     zassert_true(thread_switched_found, "doesn't find thread_switched.");
     zassert_true(sys_trace_void_found, "doesn't find sys_trace_void.");
     zassert_true(sys_trace_semaphore_found,
             "doesn't find thread_idle_entered.");
#else
     /* Should produce any tracing data when tracing disabled. */
     zassert_false(thread_switched_found, "shouldn't have trace data.");
     zassert_false(sys_trace_void_found, "shouldn't have trace data.");
     zassert_false(sys_trace_semaphore_found, "shouldn't have trace data.");
#endif
```

File: tests/subsys/tracing/CMakeLists.txt (hunk 2 above)
Comment (youhuazhu) [reply tests/subsys/tracing/CMakeLists.txt:5]: not used yet, I will remove this.  thanks .

File: tests/subsys/tracing/prj_uart_ctf.conf (hunk 3 above)
Comment (youhuazhu) [reply tests/subsys/tracing/prj_uart_ctf.conf:1]: yes, I will submit another PR for checking tracing ctf,  so will remove this file firstly.  thanks for your reminding.

File: tests/subsys/tracing/testcase.yaml (hunk 4 above)
Comment (youhuazhu) [reply tests/subsys/tracing/testcase.yaml:3]: as above reply, This is for the case when tracing functions disabled, if tracing functions are disabled, we shouldn't find any tracing data in the backend.

File: tests/subsys/tracing/testcase.yaml (hunk 6 above)
Comment (youhuazhu) [reply tests/subsys/tracing/testcase.yaml:10]: this tests worked for phisical boards,  and Let me have a double check if qemu support tracing output....

**Comment by youhuazhu** [comment]
Hi @nashif ,  Have addressed all your comments,   Any ohter comment or suggestions?  thanks very much.

File: tests/subsys/tracing/CMakeLists.txt (hunk 2 above)
Comment (youhuazhu) [reply tests/subsys/tracing/CMakeLists.txt]: Have fixed this, thanks.

File: tests/subsys/tracing/prj_uart_ctf.conf (hunk 3 above)
Comment (youhuazhu) [reply tests/subsys/tracing/prj_uart_ctf.conf]: Have removed prj_uart_ctf.conf , thanks.

File: tests/subsys/tracing/testcase.yaml (hunk 6 above)
Comment (youhuazhu) [reply tests/subsys/tracing/testcase.yaml]: I have fixed this and enabled the qemu platform.

**Comment by youhuazhu** [comment]
Hi @nashif , would you take a time to review this PR again? thanks.

File: tests/subsys/tracing/src/main.c [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+		k_sem_give(other_sem);
+	}
+}
+
+#if (defined(CONFIG_TRACING_BACKEND_UART) || defined(CONFIG_TRACING_BACKEND_USB))
+static void tracing_backends_output(
+		const struct tracing_backend *backend,
+		uint8_t *data, uint32_t length)
+{
+	uint8_t recv_char[100], recv_index = 0;
```
Comment (LeiW000) [inline tests/subsys/tracing/src/main.c:51]: I don't think you have to define a new buffer. Only using "*data" is enough.

File: tests/subsys/tracing/src/main.c [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+ * @ingroup subsys_tracing_tests
+ */
+void test_tracing_function(void)
+{
+#ifdef CONFIG_TRACING_CTF
+	ztest_test_skip();
+#endif
+	generate_more_tracing_data();
+#ifdef CONFIG_TRACING
+	zassert_true(thread_switched_found, "doesn't find thread_switched.");
```
Comment (LeiW000) [inline tests/subsys/tracing/src/main.c:122]: Its better to change the error message like below.
"thread_switched can't be found".

Same to line 123, 125, 129 and 130.

**Review by LeiW000 — CHANGES_REQUESTED** [review]

File: tests/subsys/tracing/src/tracing_string_format_test.c [hunk 9]
```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
```
[the remark this answers is not in the document]
Comment (youhuazhu) [reply tests/subsys/tracing/src/tracing_string_format_test.c]: done, thanks steven.

File: tests/subsys/tracing/src/main.c (hunk 7 above)
Comment (youhuazhu) [reply tests/subsys/tracing/src/main.c:51]: recv_char[] is used to store a whole line string,  and *data just get a char from backend,  so we need define a new buffer here.

File: tests/subsys/tracing/src/main.c (hunk 8 above)
Comment (youhuazhu) [reply tests/subsys/tracing/src/main.c]: yes,  it's more clear,  I have fixed them, thanks.

File: tests/subsys/tracing/src/main.c (hunk 7 above)
Comment (youhuazhu) [reply tests/subsys/tracing/src/main.c:51]: Let me have a try,

**Review by LeiW000 — APPROVED** [review]
It looks pefect to me now. 👍

**Comment by youhuazhu** [comment]
Thanks @LeiW000 for review,  
Hi @nashif , Can you help to review this PR again?

**Comment by youhuazhu** [comment]
Hi @nashif , Any other comments else?  Please help to review it,

**Review by mengxianglinx — APPROVED** [review]

**Review by enjiamai — APPROVED** [review]

**Comment by youhuazhu** [comment]
Hi @nashif,  Any other coomments for this PR? since this PR waiting for a long time.

**Comment by youhuazhu** [comment]
reabsed to the latest code base.  @nashif  please help to review.

File: tests/subsys/tracing/test_ctf.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    #capture trace date for 5s.
+    time.sleep(5)
+    data_path = sys.argv[1]
+    os.environ['data_path'] = str(data_path)
+
+    if not os.path.exists(data_path + "/ctf"):
+        os.mkdir(data_path + "/ctf")
+
+    if os.path.exists("subsys/tracing/ctf/tsdl/metadata"):
+        os.system('cp subsys/tracing/ctf/tsdl/metadata ${data_path}/ctf/')
```
Comment (LeiW000) [inline tests/subsys/tracing/test_ctf.py:28]: It better to use "High-level file operations" to copy file.

from shutil import copyfile
copyfile(src, dst)

File: scripts/sanity_chk/harness.py [hunk 11]
```diff
@@ -121,6 +127,29 @@ def handle(self, line):
...
 
+class Pytest(Harness):
+    def handle(self, line):
+        if self.handle_script and self.BOOTING_ZHEPHYR in line:
+            with subprocess.Popen([self.handle_script, self.running_dir], stdout=subprocess.PIPE) as proc:
+                try:
+                    stdout, _ = proc.communicate(15)
+                    if "failed" not in stdout.decode():
+                        self.state = "passed"
+                    elif "skip" in stdout.decode():
```
Comment (LeiW000) [inline scripts/sanity_chk/harness.py:138]: Please double confirm if "skip" can be shown with pytest.

**Review by LeiW000 — CHANGES_REQUESTED** [review]

File: tests/subsys/tracing/test_ctf.py (hunk 10 above)
Comment (youhuazhu) [reply tests/subsys/tracing/test_ctf.py]: Thanks Steven, Have changed  to copyfile.

File: scripts/sanity_chk/harness.py (hunk 11 above)
Comment (youhuazhu) [reply scripts/sanity_chk/harness.py:138]: Never show "skip" with pytest,  but test_ctf.py will show 'skip'  when babeltrace not installed.

**Review by LeiW000 — APPROVED** [review]

**Comment by LeiW000** [comment]
Ping @nashif to review it.

**Review by LeiW000 — APPROVED** [review]

**Comment by LeiW000** [comment]
@nashif , can it be merged now?

**Comment by youhuazhu** [comment]
Hi @nashif  Can you help to review this PR which handle tracing non-ctf case ?

**Comment by Zhaoningx** [comment]
This PR has duplicated by https://github.com/zephyrproject-rtos/zephyr/pull/31387 , because the PR's owner Youhua resigned.

**Pull request closed without merging**